yarpl/test/FlowableFlatMapTest.cpp

```diff
@@ -0,0 +1,277 @@
+/// exception was sent, the exception is thrown.
+template <typename T>
+std::vector<T> run(
+    Reference<Flowable<T>> flowable,
+    int64_t requestCount = 100) {
+  auto subscriber = make_ref<TestSubscriber<T>>(requestCount);
+  flowable->subscribe(subscriber);
+  return std::move(subscriber->values());
+}
+
+} // namespace
+
+template <typename Pred>
+std::vector<int64_t> filter_run(std::vector<int64_t> in, Pred pred) {
+  std::vector<int64_t> ret;
+  std::copy_if(in.begin(), in.end(), std::back_inserter(ret), pred);
+  return ret;
+}
+
+std::vector<int64_t>
+filter_range(std::vector<int64_t> in, int64_t startat, int64_t endat) {
+  CHECK_LE(startat, endat);
+  return filter_run(
+      in, [=](int64_t i) { return (i >= startat) && (i < endat); });
+}
+
+auto make_flowable_mapper_func() {
+  return folly::Function<Reference<Flowable<int64_t>>(int)>([](int n) {
+    switch (n) {
+      case 10:
+        return Flowables::range(n, 2);
+      case 20:
+        return Flowables::range(n, 3);
+      case 30:
+        return Flowables::range(n, 4);
+    }
+    return Flowables::range(n, 3);
+  });
+}
+
+// assumes that separate streams of values in separate_streams are entirely
+// disjoint
+template <typename T>
+bool validate_flatmapped_values(
+    std::vector<T> flatmapped,
+    std::vector<std::deque<T>> separate_streams) {
+  for (auto elem : flatmapped) {
+    bool found_match = false;
+    for (auto& stream : separate_streams) {
+      if (stream.size() > 0) {
+        if (elem == stream[0]) {
+          stream.pop_front();
+          found_match = true;
+          break;
+        }
+      }
+    }
+
+    EXPECT_TRUE(found_match)
+        << "Did not find elem '" << elem << "' in any input streams";
+    if (!found_match) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+TEST(FlowableFlatMapTest, AllRequestedTest) {
+  auto f =
+      Flowables::justN<int>({10, 20, 30})->flatMap(make_flowable_mapper_func());
+
+  std::vector<int64_t> res = run(f);
+  EXPECT_EQ(9UL, res.size());
+  EXPECT_EQ(filter_range(res, 10, 20), std::vector<int64_t>({10, 11}));
+  EXPECT_EQ(filter_range(res, 20, 30), std::vector<int64_t>({20, 21, 22}));
+  EXPECT_EQ(filter_range(res, 30, 40), std::vector<int64_t>({30, 31, 32, 33}));
+}
+
+TEST(FlowableFlatMapTest, FiniteRequested) {
+  auto f =
+      Flowables::justN<int>({10, 20, 30})->flatMap(make_flowable_mapper_func());
+
+  auto subscriber = make_ref<TestSubscriber<int64_t>>(1);
+  f->subscribe(subscriber);
+
+  EXPECT_EQ(1UL, subscriber->values().size());
+  EXPECT_TRUE(
+      validate_flatmapped_values(subscriber->values(), {{10}, {20}, {30}}));
+
+  subscriber->request(3);
+  EXPECT_TRUE(validate_flatmapped_values(
+      subscriber->values(), {{10, 11}, {20, 21, 22}, {30, 31, 32, 33}}));
+
+  subscriber->cancel();
+  EXPECT_EQ(subscriber->getValueCount(), 4);
+}
+
+TEST(FlowableFlatMapTest, MappingLambdaThrowsErrorOnFirstCall) {
```

Comment:
What if only one of the providers of the flatmap errors? Should we drop it and continue with the rest or should we error?

Can we check the case that if doesn't throw for the first call and we have multiple providers?

Reply:
I thought about this; I'm not sure if we should ignore the stream or not. @benjchristensen could you weigh in here? 

It seems like a bad idea to drop errors silently; perhaps if there's another way to supply the downstream with the errors? Eg a flatMap variant with the signature `Flowable<T> -> (T -> Flowable<R>) -> Flowable<Either<R, Error>>` ?